File: python/src/tdcv2/compute/evaluate.py
"""The evaluator for ``<compute>`` — a check digit written as a tag tree.

The tree IS the program. There is no separate expression language to parse, and no separate AST:
every implementation walks the same parse tree, so a checksum written once behaves identically in
all three. That is the whole reason this layer exists in tag form rather than as a little
expression syntax — an expression syntax would need three parsers that agree.

Pure: no random numbers, no clock, no files. The only inputs are the values a ``<field>`` can
reach, which is exactly what ``${{name}}`` can reach.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field

from ..format.mask import apply_mask
from ..format.transforms import apply_case, apply_group, apply_replace, apply_slice, apply_trim
from ..lib import numbers
from ..parser.config_builder import attributes
from .encode import encode_char
from .value import (
    ComputeError,
    Value,
    coerce_int,
    coerce_str,
    euclidean_mod,
    floor_div,
    int_value,
    list_value,
    parse_int_strict,
    str_value,
    to_output,
)

_WRAPPERS = ("result", "do", "over", "init", "in", "index", "then", "otherwise")
_CASES = ("upper", "lower", "capitalize", "title")


@dataclass(frozen=True, slots=True)
class Scope:
    """What a node can see.

    ``current``, ``current_index`` and ``acc`` are present only inside a ``<do>`` body, which is
    what makes ``<current/>`` outside an iteration an error rather than an empty string.
    """

    fields: Callable[[str], str | None]
    variables: dict[str, Value] = field(default_factory=dict)
    current: Value | None = None
    current_index: int | None = None
    acc: Value | None = None

    def with_var(self, name: str, value: Value) -> Scope:
        # The whole scope carries over: a <let> inside an iteration must not drop the contextual
        # current/current_index/acc.
        return Scope(
            self.fields,
            {**self.variables, name: value},
            self.current,
            self.current_index,
            self.acc,
        )

    def iterating(self, current: Value, index: int, acc: Value | None = None) -> Scope:
        return Scope(self.fields, self.variables, current, index, acc)


@dataclass(frozen=True, slots=True)
class Node:
    """One compute element: its tag, its attributes, and its child elements."""

    name: str
    attrs: dict[str, str]
    children: list


def evaluate(compute_element, fields: Callable[[str], str | None]) -> str:
    """A ``<compute>`` element evaluated to its output string."""
    return to_output(_slot(_children_of(compute_element), Scope(fields)))


def evaluate_predicate(valid_element, fields: Callable[[str], str | None]) -> bool:
    """A ``<valid>`` element's single predicate, for a pack generator's reject-and-retry."""
    children = _children_of(valid_element)
    if not children:
        raise ComputeError("<valid> requires a predicate child")
    return _predicate(children[0], Scope(fields))


# ── walking ─────────────────────────────────────────────────────────────────────────────────


def _node(element) -> Node:
    open_el = element.openCloseElement()
    if open_el is not None:
        return Node(open_el.name.text, attributes(open_el.attr()), _content(open_el.content()))
    self_closing = element.selfClosingElement()
    if self_closing is None:
        raise ComputeError("unexpected <data> or malformed element inside <compute>")
    return Node(self_closing.name.text, attributes(self_closing.attr()), [])


def _name_of(element) -> str:
    open_el = element.openCloseElement()
    if open_el is not None:
        return open_el.name.text
    self_closing = element.selfClosingElement()
    return self_closing.name.text if self_closing is not None else ""


def _content(content) -> list:
    return [] if content is None else list(content.element() or [])


def _children_of(element) -> list:
    return _content(element.content())


def _child(node: Node, name: str) -> Node:
    for child in node.children:
        if _name_of(child) == name:
            return _node(child)
    raise ComputeError(f"<{node.name}> requires a <{name}> child")


def _two(node: Node):
    if len(node.children) != 2:
        raise ComputeError(f"<{node.name}> requires exactly 2 children")
    return node.children[0], node.children[1]


def _slot(children: list, scope: Scope) -> Value:
    """Zero or more ``<let>`` prefixes, then exactly one value-producing expression.

    The bindings accumulate, so a later ``<let>`` and the final expression both see the earlier
    ones. That is what makes a long checksum readable: each step gets a name.
    """
    local = scope
    result: Value | None = None
    for child in children:
        if _name_of(child) == "let":
            binding = _node(child)
            local = local.with_var(binding.attrs.get("name", ""), _slot(binding.children, local))
        else:
            result = _element(child, local)
    if result is None:
        raise ComputeError("empty expression slot: no value produced")
    return result


def _wrapper(node: Node, name: str, scope: Scope) -> Value:
    return _slot(_child(node, name).children, scope)


def _iterable(value: Value) -> list[Value]:
    if value.kind == "str":
        return [str_value(ch) for ch in value.text]
    if value.kind == "list":
        return list(value.items)
    raise ComputeError("<over>: expected a string or list to iterate")


# The builtin row counters, which are numbers rather than text. _first and _last are
# deliberately absent: they are the strings "true" and "false".
_NUMERIC_BUILTIN_FIELDS = frozenset({"_count", "_total"})


def _element(element, scope: Scope) -> Value:
    n = _node(element)
    name = n.name

    # ── literals ──
    if name == "int":
        raw = n.attrs.get("v", "")
        try:
            return int_value(int(raw))
        except ValueError:
            raise ComputeError(f'<int>: "{raw}" is not an integer') from None
    if name == "str":
        return str_value(n.attrs.get("v", ""))
    if name == "list":
        raw = n.attrs.get("v")
        if raw is not None:
            items = []
            for part in (p.strip() for p in raw.split(",")):
                if not part:
                    continue
                try:
                    items.append(int_value(int(part)))
                except ValueError:
                    raise ComputeError(f'<list>: "{part}" is not an integer') from None
            return list_value(items)
        return list_value(_element(c, scope) for c in n.children)

    # ── references ──
    if name == "field":
        key = n.attrs.get("name", "")
        value = scope.fields(key)
        if value is None:
            raise ComputeError(f'<field>: "{key}" is not in scope')
        # A sequence's value is text, and coerce_int deliberately refuses a multi-digit string so
        # that "the third character" and "the number 375" stay different things. The row counters
        # are not text: _count and _total are numbers by nature. Without this they were strings,
        # so the single-digit escape hatch carried them to row 9 and the tenth row failed.
        # _first and _last stay out: they are the words "true" and "false".
        if key in _NUMERIC_BUILTIN_FIELDS:
            return int_value(int(value))
        return str_value(value)
    if name == "var":
        key = n.attrs.get("name", "")
        if key not in scope.variables:
            raise ComputeError(f'<var>: "{key}" is not bound')
        return scope.variables[key]
    if name == "current":
        if scope.current is None:
            raise ComputeError("<current/> used outside an iteration")
        return scope.current
    if name == "current_index":
        if scope.current_index is None:
            raise ComputeError("<current_index/> used outside an iteration")
        return int_value(scope.current_index)
    if name == "acc":
        if scope.acc is None:
            raise ComputeError("<acc/> used outside a <reduce>")
        return scope.acc
    if name == "let":
        raise ComputeError("<let> is a binding prefix, not a value expression")

    # ── collections ──
    if name == "each":
        items = _iterable(_wrapper(n, "over", scope))
        body = _child(n, "do")
        return list_value(
            _slot(body.children, scope.iterating(item, i)) for i, item in enumerate(items)
        )
    if name == "reduce":
        items = _iterable(_wrapper(n, "over", scope))
        body = _child(n, "do")
        acc = _wrapper(n, "init", scope)
        for i, item in enumerate(items):
            acc = _slot(body.children, scope.iterating(item, i, acc))
        return acc
    if name == "join":
        value = _slot(n.children, scope)
        if value.kind != "list":
            raise ComputeError("<join>: expected a list")
        return str_value(n.attrs.get("sep", "").join(coerce_str(v) for v in value.items))
    if name == "at":
        collection = _wrapper(n, "in", scope)
        if collection.kind != "list":
            raise ComputeError("<at>: <in> must be a list")
        index = coerce_int(_wrapper(n, "index", scope), "<at> index")
        if 0 <= index < len(collection.items):
            return collection.items[index]
        default = n.attrs.get("default")
        if default is not None:
            return int_value(parse_int_strict(default))
        raise ComputeError(f"<at>: index {index} is out of range and no default is set")
    if name == "length":
        value = _slot(n.children, scope)
        if value.kind == "str":
            return int_value(len(value.text))
        if value.kind == "list":
            return int_value(len(value.items))
        raise ComputeError("<length>: expected a string or list")

    # ── arithmetic ──
    if name == "add":
        return int_value(sum(coerce_int(_element(c, scope), "<add>") for c in n.children))
    if name == "multiply":
        product = 1
        for c in n.children:
            product *= coerce_int(_element(c, scope), "<multiply>")
        return int_value(product)
    if name == "subtract":
        if not n.children:
            raise ComputeError("<subtract> requires at least one child")
        head, *rest = n.children
        total = coerce_int(_element(head, scope), "<subtract>")
        for c in rest:
            total -= coerce_int(_element(c, scope), "<subtract>")
        return int_value(total)
    if name == "mod":
        a, b = _two(n)
        return int_value(
            euclidean_mod(coerce_int(_element(a, scope)), coerce_int(_element(b, scope)))
        )
    if name == "divide":
        a, b = _two(n)
        return int_value(floor_div(coerce_int(_element(a, scope)), coerce_int(_element(b, scope))))

    # ── encoding and conversion ──
    if name == "encode":
        value = _slot(n.children, scope)
        if value.kind != "str":
            raise ComputeError("<encode>: expected a single-character string")
        return str_value(encode_char(value.text, n.attrs.get("as", "")))
    if name == "to_number":
        return int_value(parse_int_strict(coerce_str(_slot(n.children, scope))))
    if name == "pad":
        width = int(numbers.parse(n.attrs.get("width", "0")))
        fill = n.attrs.get("fill", "0")
        return str_value(_pad_start(coerce_str(_slot(n.children, scope)), width, fill))
    if name == "concat":
        return str_value("".join(coerce_str(_element(c, scope)) for c in n.children))

    # ── text ──
    if name in _CASES:
        return str_value(apply_case(name, coerce_str(_slot(n.children, scope))))
    if name == "mask":
        pattern = n.attrs.get("pattern", "")
        return str_value(apply_mask(pattern, coerce_str(_slot(n.children, scope))))
    if name == "slice":
        to = n.attrs.get("to")
        return str_value(
            apply_slice(
                coerce_str(_slot(n.children, scope)),
                numbers.parse(n.attrs.get("from", "0")),
                None if to is None else numbers.parse(to),
            )
        )
    if name == "replace":
        return str_value(
            apply_replace(
                coerce_str(_slot(n.children, scope)),
                n.attrs.get("from", ""),
                n.attrs.get("to", ""),
            )
        )
    if name == "trim":
        return str_value(apply_trim(coerce_str(_slot(n.children, scope))))
    if name == "group":
        return str_value(
            apply_group(
                coerce_str(_slot(n.children, scope)),
                numbers.parse(n.attrs.get("size", "3")),
                n.attrs.get("sep", " "),
            )
        )

    # ── conditional ──
    if name == "choose":
        return _choose(n, scope)

    if name in _WRAPPERS:
        return _slot(n.children, scope)

    raise ComputeError(f"unknown compute tag <{name}>")


def _choose(node: Node, scope: Scope) -> Value:
    otherwise: Node | None = None
    for child in node.children:
        branch = _node(child)
        if branch.name == "when":
            if _test(_child(branch, "test"), scope):
                return _slot(_child(branch, "then").children, scope)
        elif branch.name == "otherwise":
            otherwise = branch
    if otherwise is not None:
        return _slot(otherwise.children, scope)
    raise ComputeError("<choose>: no <when> matched and no <otherwise> present")


def _test(test: Node, scope: Scope) -> bool:
    if not test.children:
        raise ComputeError("<test> requires a predicate child")
    return _predicate(test.children[0], scope)


def _predicate(element, scope: Scope) -> bool:
    n = _node(element)
    if n.name in ("equals", "greater_than", "less_than"):
        a, b = _two(n)
        x = coerce_int(_element(a, scope), f"<{n.name}>")
        y = coerce_int(_element(b, scope), f"<{n.name}>")
        if n.name == "equals":
            return x == y
        if n.name == "greater_than":
            return x > y
        return x < y
    if n.name == "is_digit":
        if not n.children:
            raise ComputeError("<is_digit> requires a child")
        value = _element(n.children[0], scope)
        return value.kind == "str" and len(value.text) == 1 and "0" <= value.text <= "9"
    raise ComputeError(f"unknown predicate <{n.name}> (valid only inside <test>)")


def _pad_start(value: str, width: int, fill: str) -> str:
    """``padStart`` as JavaScript performs it: the filler repeats and is cut, not aligned."""
    if not fill or len(value) >= width:
        return value
    needed = width - len(value)
    return (fill * (needed // len(fill) + 1))[:needed] + value

File: typescript/src/compute/evaluate.ts
/**
 * Evaluator for the `compute` declarative computation layer (spec §7).
 *
 * Walks a `<compute>` element's parse subtree directly (no separate IR — this
 * matches the rest of the codebase, which operates on ANTLR contexts) and
 * produces a `Value`. The tree IS the AST: every runtime (TS/Python/Java) walks
 * the same shape, so there is no expression grammar to re-implement.
 *
 * Pure: no RNG, clock, or I/O. The only inputs are the values reachable through
 * the `fields` resolver (the same visibility as `${{name}}` — spec §3.1).
 */

import type {
  ElementContext,
  OpenCloseElementContext,
  SelfClosingElementContext,
} from '../generated/TDCParser.js';
import type { AttrMap } from '../processor/attrs.js';
import { contentElements, elementKind, elementName, extractAttrs } from '../processor/walk.js';

import {
  applyCase,
  applyGroup,
  applyMask,
  applyReplace,
  applySlice,
  applyTrim,
} from '../format/transforms.js';

import { encodeChar } from './encode.js';
import {
  coerceInt,
  coerceStr,
  ComputeError,
  euclideanMod,
  floorDiv,
  int,
  list,
  parseIntStrict,
  str,
  valueToOutput,
  type Value,
} from './value.js';

/**
 * The builtin row counters, which are numbers rather than text.
 *
 * `_first` and `_last` are deliberately absent: they are the strings "true" and
 * "false", and reading them as integers would be a lie of a different kind.
 */
const NUMERIC_BUILTIN_FIELDS: ReadonlySet<string> = new Set(['_count', '_total']);

/**
 * Evaluation scope. `fields` resolves `<field>`; `vars` holds `<let>` bindings;
 * `current`/`currentIndex`/`acc` are the contextual iteration values, present
 * only inside a `<do>` body.
 */
export interface EvalScope {
  readonly fields: (name: string) => string | undefined;
  readonly vars: ReadonlyMap<string, Value>;
  readonly current?: Value;
  readonly currentIndex?: bigint;
  readonly acc?: Value;
}

/** Normalised view of a compute element: name, attributes, child elements. */
interface CNode {
  readonly name: string;
  readonly attrs: AttrMap;
  readonly children: readonly ElementContext[];
}

function asCNode(el: ElementContext): CNode {
  const k = elementKind(el);
  if (!k || k.kind === 'data') {
    throw new ComputeError('unexpected <data> or malformed element inside <compute>');
  }
  const node: OpenCloseElementContext | SelfClosingElementContext = k.node;
  const children = k.kind === 'open' ? contentElements(k.node.content()) : [];
  return { name: elementName(node), attrs: extractAttrs(node.attr()), children };
}

function nodeName(el: ElementContext): string {
  const k = elementKind(el);
  return k && k.kind !== 'data' ? elementName(k.node) : '';
}

function childByName(n: CNode, name: string): ElementContext | undefined {
  return n.children.find((c) => nodeName(c) === name);
}

function requireChild(n: CNode, name: string): CNode {
  const child = childByName(n, name);
  if (!child) throw new ComputeError(`<${n.name}> requires a <${name}> child`);
  return asCNode(child);
}

// --- scope helpers (kept explicit for exactOptionalPropertyTypes) ---

function withVar(scope: EvalScope, name: string, value: Value): EvalScope {
  const vars = new Map(scope.vars);
  vars.set(name, value);
  // Spread the whole scope so a <let> inside an iteration keeps the contextual
  // current/currentIndex/acc — binding a variable must not drop them.
  return { ...scope, vars };
}

function withIter(scope: EvalScope, current: Value, currentIndex: bigint, acc?: Value): EvalScope {
  const base = { fields: scope.fields, vars: scope.vars, current, currentIndex };
  return acc === undefined ? base : { ...base, acc };
}

/** Turn a collection value into the list of elements iteration walks. */
function iterableOf(value: Value): Value[] {
  if (value.t === 'str') return Array.from(value.v).map((ch) => str(ch));
  if (value.t === 'list') return [...value.v];
  throw new ComputeError('<over>: expected a string or list to iterate');
}

/**
 * Evaluate a single-expression slot: zero or more `<let>` prefixes followed by
 * exactly one value-producing expression (spec §7.2). Bindings accumulate so a
 * later `<let>` and the final expression see the earlier ones.
 */
function evalSlot(children: readonly ElementContext[], scope: EvalScope): Value {
  let localScope = scope;
  let result: Value | undefined;
  for (const child of children) {
    if (nodeName(child) === 'let') {
      const cn = asCNode(child);
      const name = cn.attrs['name'] ?? '';
      localScope = withVar(localScope, name, evalSlot(cn.children, localScope));
    } else {
      result = evalElement(child, localScope);
    }
  }
  if (result === undefined) throw new ComputeError('empty expression slot: no value produced');
  return result;
}

/** Evaluate a named role-wrapper child (`<over>`, `<init>`, …) as a slot. */
function evalWrapper(n: CNode, wrapper: string, scope: EvalScope): Value {
  return evalSlot(requireChild(n, wrapper).children, scope);
}

function evalElement(el: ElementContext, scope: EvalScope): Value {
  const n = asCNode(el);
  switch (n.name) {
    // --- literals ---
    case 'int': {
      const raw = n.attrs['v'] ?? '';
      if (!/^-?[0-9]+$/.test(raw)) throw new ComputeError(`<int>: "${raw}" is not an integer`);
      return int(BigInt(raw));
    }
    case 'str':
      return str(n.attrs['v'] ?? '');
    case 'list': {
      const raw = n.attrs['v'];
      if (raw !== undefined) {
        const parts = raw
          .split(',')
          .map((s) => s.trim())
          .filter((s) => s.length > 0);
        return list(
          parts.map((p) => {
            if (!/^-?[0-9]+$/.test(p)) throw new ComputeError(`<list>: "${p}" is not an integer`);
            return int(BigInt(p));
          }),
        );
      }
      return list(n.children.map((c) => evalElement(c, scope)));
    }

    // --- references ---
    case 'field': {
      const name = n.attrs['name'] ?? '';
      const value = scope.fields(name);
      if (value === undefined) throw new ComputeError(`<field>: "${name}" is not in scope`);
      // A sequence's value is text, and `coerceInt` deliberately refuses a
      // multi-digit string so that "the third character" and "the number 375"
      // stay different things. The row counters are not text: `_count` and
      // `_total` are numbers by nature, and typing them as such is what makes
      // `<mod><field name="_count"/><int v="2"/></mod>` mean something.
      //
      // Without this the counters were strings, so the single-digit escape
      // hatch in `coerceInt` carried them to row 9 and the tenth row failed
      // with "expected an integer in arithmetic, got the string "10"". A
      // config tested on five rows broke on a million, which is the worst
      // shape a bug can have here.
      if (NUMERIC_BUILTIN_FIELDS.has(name)) return int(BigInt(value));
      return str(value);
    }
    case 'var': {
      const name = n.attrs['name'] ?? '';
      const value = scope.vars.get(name);
      if (value === undefined) throw new ComputeError(`<var>: "${name}" is not bound`);
      return value;
    }
    case 'current':
      if (scope.current === undefined) {
        throw new ComputeError('<current/> used outside an iteration');
      }
      return scope.current;
    case 'current_index':
      if (scope.currentIndex === undefined) {
        throw new ComputeError('<current_index/> used outside an iteration');
      }
      return int(scope.currentIndex);
    case 'acc':
      if (scope.acc === undefined) {
        throw new ComputeError('<acc/> used outside a <reduce>');
      }
      return scope.acc;
    case 'let':
      throw new ComputeError('<let> is a binding prefix, not a value expression');

    // --- collections ---
    case 'each': {
      const items = iterableOf(evalWrapper(n, 'over', scope));
      const doNode = requireChild(n, 'do');
      return list(
        items.map((item, i) => evalSlot(doNode.children, withIter(scope, item, BigInt(i)))),
      );
    }
    case 'reduce': {
      const items = iterableOf(evalWrapper(n, 'over', scope));
      const doNode = requireChild(n, 'do');
      let acc = evalWrapper(n, 'init', scope);
      items.forEach((item, i) => {
        acc = evalSlot(doNode.children, withIter(scope, item, BigInt(i), acc));
      });
      return acc;
    }
    case 'join': {
      const sep = n.attrs['sep'] ?? '';
      const value = evalSlot(n.children, scope);
      if (value.t !== 'list') throw new ComputeError('<join>: expected a list');
      return str(value.v.map(coerceStr).join(sep));
    }
    // The exact inverse of <join>, and the fourth way to get a list.
    //
    // Before it there were three — a literal <list v="…">, the result of <each>, and a string
    // walked CHARACTER by character — and none of them could read back a column that `repeat=`
    // had joined. So "sum quantity × price over the lines of this order" could not be said at
    // all unless the two lists happened to have the same length.
    //
    // The pieces are STRINGS, like every other piece of text in this layer: <to_number> turns
    // one into an integer, and the arithmetic tags say so when they are handed text.
    case 'split': {
      const sep = n.attrs['sep'] ?? '';
      // An empty separator is refused rather than given a meaning. JavaScript would answer with
      // every character, Python would refuse outright, and the other three differ again — so any
      // reading here would make one implementation disagree with the rest. Walking a string
      // character by character already has a spelling: <over> takes a string directly.
      if (sep === '') {
        throw new ComputeError(
          '<split>: sep= is empty — to walk a string character by character, put it in <over> ' +
            'directly, which is what an empty separator would have to mean',
        );
      }
      const value = evalSlot(n.children, scope);
      if (value.t !== 'str') {
        throw new ComputeError('<split>: expected a string, got a list');
      }
      return list(value.v.split(sep).map((piece) => str(piece)));
    }
    case 'at': {
      const coll = evalWrapper(n, 'in', scope);
      if (coll.t !== 'list') throw new ComputeError('<at>: <in> must be a list');
      const idx = Number(coerceInt(evalWrapper(n, 'index', scope), '<at> index'));
      const found = coll.v[idx];
      if (found !== undefined) return found;
      const dflt = n.attrs['default'];
      if (dflt !== undefined) return int(parseIntStrict(dflt));
      throw new ComputeError(`<at>: index ${String(idx)} is out of range and no default is set`);
    }
    case 'length': {
      const value = evalSlot(n.children, scope);
      if (value.t === 'str') return int(BigInt(Array.from(value.v).length));
      if (value.t === 'list') return int(BigInt(value.v.length));
      throw new ComputeError('<length>: expected a string or list');
    }

    // --- arithmetic ---
    case 'add': {
      let sum = 0n;
      for (const c of n.children) sum += coerceInt(evalElement(c, scope), '<add>');
      return int(sum);
    }
    case 'multiply': {
      let product = 1n;
      for (const c of n.children) product *= coerceInt(evalElement(c, scope), '<multiply>');
      return int(product);
    }
    case 'subtract': {
      const [head, ...rest] = n.children;
      if (head === undefined) throw new ComputeError('<subtract> requires at least one child');
      let acc = coerceInt(evalElement(head, scope), '<subtract>');
      for (const c of rest) acc -= coerceInt(evalElement(c, scope), '<subtract>');
      return int(acc);
    }
    case 'mod': {
      const [a, b] = requireTwo(n);
      return int(euclideanMod(coerceInt(evalElement(a, scope)), coerceInt(evalElement(b, scope))));
    }
    case 'divide': {
      const [a, b] = requireTwo(n);
      return int(floorDiv(coerceInt(evalElement(a, scope)), coerceInt(evalElement(b, scope))));
    }

    // --- encoding / conversion ---
    case 'encode': {
      const as = n.attrs['as'] ?? '';
      const value = evalSlot(n.children, scope);
      if (value.t !== 'str') throw new ComputeError('<encode>: expected a single-character string');
      return str(encodeChar(value.v, as));
    }
    case 'to_number':
      return int(parseIntStrict(coerceStr(evalSlot(n.children, scope))));
    case 'pad': {
      const width = Number(n.attrs['width'] ?? '0');
      const fill = n.attrs['fill'] ?? '0';
      return str(coerceStr(evalSlot(n.children, scope)).padStart(width, fill));
    }
    case 'concat':
      return str(n.children.map((c) => coerceStr(evalElement(c, scope))).join(''));

    // --- text formatting ---
    case 'upper':
    case 'lower':
    case 'capitalize':
    case 'title':
      return str(applyCase(n.name, coerceStr(evalSlot(n.children, scope))));
    case 'mask':
      return str(applyMask(n.attrs['pattern'] ?? '', coerceStr(evalSlot(n.children, scope))));
    case 'slice': {
      const to = n.attrs['to'];
      return str(
        applySlice(
          coerceStr(evalSlot(n.children, scope)),
          Number(n.attrs['from'] ?? '0'),
          to === undefined ? undefined : Number(to),
        ),
      );
    }
    case 'replace':
      return str(
        applyReplace(
          coerceStr(evalSlot(n.children, scope)),
          n.attrs['from'] ?? '',
          n.attrs['to'] ?? '',
        ),
      );
    case 'trim':
      return str(applyTrim(coerceStr(evalSlot(n.children, scope))));
    case 'group':
      return str(
        applyGroup(
          coerceStr(evalSlot(n.children, scope)),
          Number(n.attrs['size'] ?? '3'),
          n.attrs['sep'] ?? ' ',
        ),
      );

    // --- conditional ---
    case 'choose':
      return evalChoose(n, scope);

    // --- role wrappers evaluated as slots ---
    case 'result':
    case 'do':
    case 'over':
    case 'init':
    case 'in':
    case 'index':
    case 'then':
    case 'otherwise':
      return evalSlot(n.children, scope);

    default:
      throw new ComputeError(`unknown compute tag <${n.name}>`);
  }
}

function requireTwo(n: CNode): [ElementContext, ElementContext] {
  const [a, b] = n.children;
  if (a === undefined || b === undefined || n.children.length !== 2) {
    throw new ComputeError(`<${n.name}> requires exactly 2 children`);
  }
  return [a, b];
}

function evalChoose(n: CNode, scope: EvalScope): Value {
  let otherwise: CNode | undefined;
  for (const child of n.children) {
    const cn = asCNode(child);
    if (cn.name === 'when') {
      if (evalTest(requireChild(cn, 'test'), scope)) {
        return evalSlot(requireChild(cn, 'then').children, scope);
      }
    } else if (cn.name === 'otherwise') {
      otherwise = cn;
    }
  }
  if (otherwise) return evalSlot(otherwise.children, scope);
  throw new ComputeError('<choose>: no <when> matched and no <otherwise> present');
}

/** Evaluate a `<test>` (its single predicate child) to a boolean. */
function evalTest(test: CNode, scope: EvalScope): boolean {
  const pred = test.children[0];
  if (pred === undefined) throw new ComputeError('<test> requires a predicate child');
  return evalPredicate(pred, scope);
}

function evalPredicate(el: ElementContext, scope: EvalScope): boolean {
  const n = asCNode(el);
  switch (n.name) {
    case 'equals':
    case 'greater_than':
    case 'less_than': {
      const [a, b] = requireTwo(n);
      const x = coerceInt(evalElement(a, scope), `<${n.name}>`);
      const y = coerceInt(evalElement(b, scope), `<${n.name}>`);
      if (n.name === 'equals') return x === y;
      if (n.name === 'greater_than') return x > y;
      return x < y;
    }
    case 'is_digit': {
      const child = n.children[0];
      if (child === undefined) throw new ComputeError('<is_digit> requires a child');
      const value = evalElement(child, scope);
      return value.t === 'str' && /^[0-9]$/.test(value.v);
    }
    default:
      throw new ComputeError(`unknown predicate <${n.name}> (valid only inside <test>)`);
  }
}

/**
 * Evaluate a `<compute>` element to its output string. `fields` resolves
 * `<field name>` references to their in-scope values.
 */
export function evaluateCompute(
  computeEl: OpenCloseElementContext,
  fields: (name: string) => string | undefined,
): string {
  const scope: EvalScope = { fields, vars: new Map() };
  return valueToOutput(evalSlot(contentElements(computeEl.content()), scope));
}

/**
 * Evaluate a `<valid>` element's single predicate child to a boolean. Used by
 * pack generators for reject-and-retry (migration spec §4.2).
 */
export function evaluateComputePredicate(
  validEl: OpenCloseElementContext,
  fields: (name: string) => string | undefined,
): boolean {
  const pred = contentElements(validEl.content()).find((c) => nodeName(c) !== '');
  if (pred === undefined) throw new ComputeError('<valid> requires a predicate child');
  return evalPredicate(pred, { fields, vars: new Map() });
}
